source/plugins/ruby/in_kube_pvinventory.rb

```diff
@@ -0,0 +1,253 @@
+          @@pvTelemetryTimeTracker = DateTime.now.to_time.to_i
+        end
+
+      rescue => errorStr
+        $log.warn "in_kube_pvinventory::enumerate:Failed in enumerate: #{errorStr}"
+        $log.debug_backtrace(errorStr.backtrace)
+        ApplicationInsightsUtility.sendExceptionTelemetry(errorStr)
+      end
+    end # end enumerate
+
+    def parse_and_emit_records(pvInventory, batchTime = Time.utc.iso8601)
+      currentTime = Time.now
+      emitTime = currentTime.to_f
+      eventStream = MultiEventStream.new
+
+      begin
+        records = []
+        pvInventory["items"].each do |item|
+
+          # Node, pod, & usage info can be found by joining with pvUsedBytes metric using PVCNamespace/PVCName
+          record = {}
+          record["CollectionTime"] = batchTime
+          record["ClusterId"] = KubernetesApiClient.getClusterId
+          record["ClusterName"] = KubernetesApiClient.getClusterName
+          record["PVName"] = item["metadata"]["name"]
+          record["PVStatus"] = item["status"]["phase"]
+          record["PVAccessModes"] = item["spec"]["accessModes"].join(', ')
+          record["PVStorageClassName"] = item["spec"]["storageClassName"]
+          record["PVCapacityBytes"] = KubernetesApiClient.getMetricNumericValue("memory", item["spec"]["capacity"]["storage"])
+          record["PVCreationTimeStamp"] = item["metadata"]["creationTimestamp"]
+
+          # Optional values
+          pvcNamespace, pvcName = getPVCInfo(item)
+          type, typeInfo = getTypeInfo(item)
+          record["PVCNamespace"] = pvcNamespace
+          record["PVCName"] = pvcName
+          record["PVType"] = type
+          record["PVTypeInfo"] = typeInfo
+
+          records.push(record)
+
+          # Record telemetry
+          if type == nil
+            type = "empty"
+          end
+          if (@pvTypeToCountHash.has_key? type)
+            @pvTypeToCountHash[type] += 1
+          else
+            @pvTypeToCountHash[type] = 1
+          end
+        end
+
+        records.each do |record|
+          if !record.nil?
+            wrapper = {
+              "DataType" => "KUBE_PV_INVENTORY_BLOB",
+              "IPName" => "ContainerInsights",
+              "DataItems" => [record.each { |k, v| record[k] = v }],
+            }
+            eventStream.add(emitTime, wrapper) if wrapper
+          end
+        end
+
+        router.emit_stream(@tag, eventStream) if eventStream
+
+      rescue => errorStr
+        $log.warn "Failed in parse_and_emit_record for in_kube_pvinventory: #{errorStr}"
+        $log.debug_backtrace(errorStr.backtrace)
+        ApplicationInsightsUtility.sendExceptionTelemetry(errorStr)
+      end
+    end
+
+    def getPVCInfo(item)
+      begin
+        if !item["spec"].nil? && !item["spec"]["claimRef"].nil?
+          claimRef = item["spec"]["claimRef"]
+          pvcNamespace = claimRef["namespace"]
```

Comment:
claimRef always has name & namespace, so nil/empty check not needed here ?

Reply:
claimRef will always have a name & namespace, but if a PV doesn't have a claim, I think there is no claimRef field at all for spec and item["spec"]["claimRef"] would be nil. I can double check.

Reply:
PV spec with no claim is below...so we have to check if claimRef exists to know if there is a PVC. Or am I misunderstanding what you meant?
![image](https://user-images.githubusercontent.com/21349004/95888671-3c89f000-0d36-11eb-8afe-576cffaaa34a.png)